### PR TITLE
Adding Radeon support by controlling wave size

### PR DESCRIPTION
**Pull request by seanofthemillers**
# Summary

This is a feature. It adds a configuration time control parameter for the default wave size. On AMD MI cards this is generally 64, however on Radeon (gaming) cards this is usually 32. These changes will require the user to know if the card is setup for Wave32 or Wave64.

I also added a fix for the dynamically sized memory allocation which seems to trigger a lot of warnings for ROCm 6.2.

# Design review (for API changes or additions---delete if unneeded)

On (date), we reviewed this PR.  We discussed the design ideas:

1. First idea or goal
2. Second idea
3. Third idea

This PR implements 1. and 3.  It leaves out 2. for the following reasons

- (impractical)
- (too big)
- (not a good idea anyway)

### CODE REVIEW

File: include/RAJA/policy/tensor/arch/hip/hip_wave.hpp [hunk 1]
```diff
@@ -780,8 +780,8 @@ namespace expt
 
         // Third: mask off everything but output_segment
         //        this is because all output segments are valid at this point
-        // (5-segbits), the 5 is since the warp-width is 32 == 1<<5
-        int our_output_segment = get_lane()>>(6-segbits);
+        const int log2_warp_size = 32-1-__builtin_clz(warpSize);
```
Comment (MrBurmark) [inline include/RAJA/policy/tensor/arch/hip/hip_wave.hpp:783]: By using warpSize it appear that this is making a runtime value now instead of a compile time value, do we need to worry about the performance implications of this?

Comment (seanofthemillers) [reply include/RAJA/policy/tensor/arch/hip/hip_wave.hpp:783]: It will add a few instructions and possibly a sgpr load. It probably won't be noticeable, but we can move the log2_warp_size as a constexpr member of the DeviceConstants. Or just wrap these two places in ifdefs. If we keep it dynamic then it should work for Nvidia/other as well - I'm not sure how important that is.

Comment (MrBurmark) [reply include/RAJA/policy/tensor/arch/hip/hip_wave.hpp:783]: Given that we're already relying on getting the right warpSize at configure time I think we should keep this compile time if possible. There is a constexpr RAJA::log2 that we can use instead of clz here to keep it constexpr.

Comment (seanofthemillers) [reply include/RAJA/policy/tensor/arch/hip/hip_wave.hpp]: I finally had a chance to update this. I switched it over to the DeviceConstants/WARP_SIZE value using the RAJA::log2 call.

File: include/RAJA/pattern/kernel/InitLocalMem.hpp [hunk 2]
```diff
@@ -77,23 +77,16 @@ struct StatementExecutor<statement::InitLocalMem<RAJA::cpu_tile_mem,camp::idx_se
...
 
     // Initialize memory
-#ifdef RAJA_COMPILER_MSVC
-    // MSVC doesn't like taking a pointer to stack allocated data?!?!
     varType *ptr = new varType[camp::get<Pos>(data.param_tuple).size()];
     camp::get<Pos>(data.param_tuple).set_data(ptr);
-#else
-    varType Array[camp::get<Pos>(data.param_tuple).size()];
-    camp::get<Pos>(data.param_tuple).set_data(&Array[0]);
-#endif
```
Comment (MrBurmark) [inline include/RAJA/pattern/kernel/InitLocalMem.hpp:87]: Is this related to this PR?

File: CMakeLists.txt [hunk 3]
```diff
@@ -171,6 +171,7 @@ endif()
 
 if(RAJA_ENABLE_HIP)
   message(STATUS "HIP version: ${hip_VERSION}")
+  set(RAJA_DEFAULT_WAVESIZE "64" CACHE STRING "Default wave size for GPU architecture. E.g. MI200/MI300 this is 64.")
```
Comment (MrBurmark) [inline CMakeLists.txt:174]: This should probably have HIP in the name.

File: include/RAJA/policy/tensor/arch/hip/hip_wave.hpp [hunk 4]
```diff
@@ -780,8 +780,8 @@ namespace expt
 
         // Third: mask off everything but output_segment
         //        this is because all output segments are valid at this point
-        // (5-segbits), the 5 is since the warp-width is 32 == 1<<5
-        int our_output_segment = get_lane()>>(6-segbits);
+        const int log2_warp_size = RAJA::log2(RAJA::policy::hip::device_constants.WARP_SIZE);
```
Comment (MrBurmark) [inline include/RAJA/policy/tensor/arch/hip/hip_wave.hpp:783]:
```suggestion
        constexpr int log2_warp_size = RAJA::log2(RAJA::policy::hip::device_constants.WARP_SIZE);
```

File: CMakeLists.txt (hunk 3 above)
Comment (seanofthemillers) [reply CMakeLists.txt]: Switched it to RAJA_HIP_WAVESIZE - I'm not sure 'default' was needed.

File: include/RAJA/policy/tensor/arch/hip/hip_wave.hpp (hunk 4 above)
Comment (seanofthemillers) [reply include/RAJA/policy/tensor/arch/hip/hip_wave.hpp]: Fixed in both places.

File: include/RAJA/pattern/kernel/InitLocalMem.hpp (hunk 2 above)
Comment (seanofthemillers) [reply include/RAJA/pattern/kernel/InitLocalMem.hpp:87]: This was to address a warning that pops up in clang 18 for dynamic stack variables. I can move to a separate issue, but it was small enough I thought I would just throw it in here.

Comment (MrBurmark) [reply include/RAJA/pattern/kernel/InitLocalMem.hpp:87]: I was wondering if anyone knew if we'd gotten past the MSVC problem from the comment.

Comment (seanofthemillers) [reply include/RAJA/pattern/kernel/InitLocalMem.hpp:87]: It would appear that the compiler is not treating it as compile time knowledge - I'm not sure why. Should I remove this change from this branch to be worked on separately?

Comment (MrBurmark) [reply include/RAJA/pattern/kernel/InitLocalMem.hpp:87]: Let's do that for now.

File: include/RAJA/policy/tensor/arch/hip/hip_wave.hpp [hunk 5]
```diff
@@ -780,8 +780,8 @@ namespace expt
 
         // Third: mask off everything but output_segment
         //        this is because all output segments are valid at this point
-        // (5-segbits), the 5 is since the warp-width is 32 == 1<<5
-        int our_output_segment = get_lane()>>(6-segbits);
+        constexpr int log2_warp_size = RAJA::log2(RAJA::policy::hip::device_constants.WARP_SIZE);
```
Comment (MrBurmark) [inline include/RAJA/policy/tensor/arch/hip/hip_wave.hpp:783]:
```suggestion
        static constexpr int log2_warp_size = RAJA::log2(RAJA::policy::hip::device_constants.WARP_SIZE);
```

File: include/RAJA/policy/tensor/arch/hip/hip_wave.hpp [hunk 6]
```diff
@@ -828,8 +828,9 @@ namespace expt
 
         // First: tree reduce values within each segment
         element_type x = m_value;
+        constexpr int log2_warp_size = RAJA::log2(RAJA::policy::hip::device_constants.WARP_SIZE);
```
Comment (MrBurmark) [inline include/RAJA/policy/tensor/arch/hip/hip_wave.hpp:831]:
```suggestion
        static constexpr int log2_warp_size = RAJA::log2(RAJA::policy::hip::device_constants.WARP_SIZE);
```

File: test/include/RAJA_test-tensor.hpp [hunk 7]
```diff
@@ -87,7 +87,9 @@ struct TensorTestHelper<RAJA::expt::hip_wave_register>
     void exec(BODY const &body){
       hipDeviceSynchronize();
 
-      RAJA::forall<RAJA::hip_exec<64>>(RAJA::RangeSegment(0,64),
+      constexpr int warp_size = RAJA::policy::hip::device_constants.WARP_SIZE;
```
Comment (MrBurmark) [inline test/include/RAJA_test-tensor.hpp:90]:
```suggestion
      static constexpr int warp_size = RAJA::policy::hip::device_constants.WARP_SIZE;
```

**Review by MrBurmark — APPROVED** [review]

**Review by rhornung67 — APPROVED** [review]
